src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/ScmCache.java

```diff
@@ -0,0 +1,106 @@
+package pl.allegro.tech.build.axion.release.infrastructure.git;
+
+import groovy.lang.Tuple2;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.revwalk.RevWalk;
+import pl.allegro.tech.build.axion.release.domain.scm.ScmException;
+import pl.allegro.tech.build.axion.release.domain.scm.ScmRepository;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides cached version for some operations on {@link ScmRepository}
+ */
+public class ScmCache {
```

Comment:
what do you think about enum based singletons? also it would be nice if this class would be package-private ;)

Reply:
It can not be easily package private right now as it is accessed from VersionResolver. Do you have suggestion how to move necessary classes so ti encapsulates better?